Unit test tries to write to installed library directory

This is a re-incarnation of #7201: 

One `astropy.io.fits` unittest tries to write to a file in the installed library. This fails if one tries to run the test suite without write access to them, f.e. as an ordinary user for a system wide installation.

This makes the [continuous integration test](https://ci.debian.net/packages/a/astropy/unstable/amd64/) of Astropy in Debian fail. From the [test log](https://ci.debian.net/data/autopkgtest/unstable/amd64/a/astropy/313486/log.gz):
```
___________________ TestFileFunctions.test_open_file_handle ____________________

self = <astropy.io.fits.tests.test_core.TestFileFunctions object at 0x7f5ef3304b00>

    def test_open_file_handle(self):
        # Make sure we can open a FITS file from an open file handle
        with open(self.data('test0.fits'), 'rb') as handle:
            with fits.open(handle) as fitsfile:
                pass
    
        with open(self.temp('temp.fits'), 'wb') as handle:
            with fits.open(handle, mode='ostream') as fitsfile:
                pass
    
        # Opening without explicitly specifying binary mode should fail
        with pytest.raises(ValueError):
            with open(self.data('test0.fits')) as handle:
                with fits.open(handle) as fitsfile:
                    pass
    
        # All of these read modes should fail
        for mode in ['r', 'rt', 'r+', 'rt+', 'a', 'at', 'a+', 'at+']:
            with pytest.raises(ValueError):
>               with open(self.data('test0.fits'), mode=mode) as handle:
E               PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-packages/astropy/io/fits/tests/data/test0.fits'

/usr/lib/python3/dist-packages/astropy/io/fits/tests/test_core.py:613: PermissionError
```
Pinging @mangalagb (who was so keen to fix #7201) to have a look here as well?

--- a/astropy/io/fits/tests/test_convenience.py
+++ b/astropy/io/fits/tests/test_convenience.py
@@ -2,7 +2,6 @@
 
 
 import os
-import shutil
 import warnings
 
 import pytest
@@ -146,14 +145,12 @@ class TestConvenience(FitsTestCase):
         Regression test for https://github.com/astropy/astropy/issues/6937
         """
         # copy fits file to the temp directory
-        filename = self.data('tb.fits')
-        temp_filename = self.temp('tb.fits')
-        shutil.copyfile(filename, temp_filename)
+        self.copy_file('tb.fits')
 
         # test without datafile
-        fits.tabledump(temp_filename)
+        fits.tabledump(self.temp('tb.fits'))
         assert os.path.isfile(self.temp('tb_1.txt'))
 
         # test with datafile
-        fits.tabledump(temp_filename, datafile=self.temp('test_tb.txt'))
+        fits.tabledump(self.temp('tb.fits'), datafile=self.temp('test_tb.txt'))
         assert os.path.isfile(self.temp('test_tb.txt'))

--- a/astropy/io/fits/tests/test_core.py
+++ b/astropy/io/fits/tests/test_core.py
@@ -608,14 +608,15 @@ class TestFileFunctions(FitsTestCase):
                     pass
 
         # All of these read modes should fail
-        for mode in ['r', 'rt', 'r+', 'rt+', 'a', 'at', 'a+', 'at+']:
+        for mode in ['r', 'rt']:
             with pytest.raises(ValueError):
                 with open(self.data('test0.fits'), mode=mode) as handle:
                     with fits.open(handle) as fitsfile:
                         pass
 
-        # These write modes should fail as well
-        for mode in ['w', 'wt', 'w+', 'wt+']:
+        # These update or write modes should fail as well
+        for mode in ['w', 'wt', 'w+', 'wt+', 'r+', 'rt+',
+                     'a', 'at', 'a+', 'at+']:
             with pytest.raises(ValueError):
                 with open(self.temp('temp.fits'), mode=mode) as handle:
                     with fits.open(handle) as fitsfile:

